src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -129,12 +132,52 @@ void FireEventsForDocument(DocumentId documentId)
             if (filename == null)
                 return;
 
-            if (!projectCodeModel.TryGetCachedFileCodeModel(filename, out var fileCodeModelHandle))
+            projectCodeModelAndFileNames.Add((projectCodeModel, filename));
+        }
+    }
+
+    private async Task FireEventsForChangedDocumentsAsync(
+        ArrayBuilder<(ProjectCodeModel projectCodeModel, string filename)> projectCodeModelAndFileNames,
+        CancellationToken cancellationToken)
+    {
+        // This logic preserves the previous behavior we had with IForegroundNotificationService.
+        // Specifically, we don't run on the UI thread for more than 15ms at a time. And we don't
+        // check the input queue more than once per ms. Once we have run for more than 15 ms,
+        // we wait 50ms before continuing.  These constants are just what we defined from
+        // legacy, and otherwise have no special meaning.
+        const int MaxTimeSlice = 15;
+        double nextInputCheckElapsedMs = 1;
+
+        var stopwatch = SharedStopwatch.StartNew();
+        foreach (var (projectCodeModel, filename) in projectCodeModelAndFileNames)
+        {
+            // If we've been asked to shutdown, don't bother reporting any more events.
+            if (_threadingContext.DisposalToken.IsCancellationRequested)
```

Comment:
should just use the CT passed in afaict.  it is already controlled by the disposal token.

Reply:
makes sense